### DISCUSSION TITLE
Download file in Livewire

### DISCUSSION THREAD

**Discussion by gid-m** [spatie/simple-excel#101] [Q&A] [answered]
Any leads on how to download the generated file on Laravel Livewire. I have followed the documentation as follows: 
`$writer = SimpleExcelWriter::streamDownload('some-name.csv');`

`$row = WriterEntityFactory::createRowFromArray(['Hello', 'World']);`

`$writer->addRow($row);`

`$writer->toBrowser();`

but nothing happens when on click of a button, I call a function containing this code.

**Comment by freekmurze** [comment]
I haven't had to do this myself yet. If you find out, let us know, and I'll put it in the readme.

**Reply by Ceepster14** [reply]
Hi,

I think it's definitely worth adding something in the readme about this as I just hit the same problem by using this package within a Livewire component and this answer saved me a lot of head scratching as otherwise the attempted download just fails silently which makes it tough to identify the cause.

**Reply by gauravgargcs** [reply]
hello can you help me in this, i am stuck also

**Comment by gid-m** [comment, marked as answer]
After tinkering with it, I ended up having to do this to start the download:
```
$writer->addRow($row);

return response()->streamDownload(function () use ($writer) {
            $writer->close();
        },  $fileName);
```
Hence omitting `$writer->toBrowser();`

**Comment by gauravgargcs** [comment]
hello, i have same issue and i can not resolve, can you please tell me what is wrong i am doing-
 $rows = [];
        User::query()
            ->where('role', 'like', '%' . $this->search . '%')
            ->where('name', 'like', '%' . $this->search_name . '%')
            ->lazyById(2000, 'id')
            ->each(function ($user) use (&$rows) {
                $rows[] = $user->toArray();
            });
            $fileName='users.csv';
            $writer= SimpleExcelWriter::streamDownload($fileName);
            $writer->addRows($rows);
            return response()->streamDownload(function () use ($writer) {
                $writer->close();
            },  $fileName);

**Reply by shokanshi** [reply]
If you are calling your code from another class, you will need to return the method call from the component too.

e.g.

```
class A {
  public function data() {
    // do something here
    return response()->streamDownload(function () use ($writer) {
        $writer->close();
    }, $fileName);
  }
}
```

```
class LiveWireComponent extends Component
{
    public function download() {
        return app(A::class)->data();
    }
}
```

**Reply by gauravgargcs** [reply]
hello, thanks for reply, i am calling in component like this, can you guide please where i am doing wrong
```
class UserManagement extends Component
{
public function export()
    {
        $rows = [];
        User::query()
            ->where('role', 'like', '%' . $this->search . '%')
            ->where('name', 'like', '%' . $this->search_name . '%')
            ->lazyById(2000, 'id')
            ->each(function ($user) use (&$rows) {
                $rows[] = $user->toArray();
            });
            $fileName='users.csv';
            $writer= SimpleExcelWriter::streamDownload($fileName);
            $writer->addRows($rows);
            return response()->streamDownload(function () use ($writer) {
                $writer->close();
            },  $fileName);
    }
}
```

**Reply by shokanshi** [reply]
You can try:

```
public function export()
{
        $fileName = 'example.xlsx';

        $writer = SimpleExcelWriter::streamDownload($fileName)
        ->nameCurrentSheet('Users')
        ->addHeader(['name', 'role']);

        User::query()->where('role', 'like', "%{$this->search}%")
        ->select(['name', 'role'])
        ->chunk(2000, function (Collection $users) use ($writer) {
                foreach ($users as $user) {
                        $writer->addRow($user->toArray());
                }

                flush();
        });

        return response()->streamDownload(function () use ($writer) {
                $writer->close();
        },  $fileName);
}
```

**Reply by gauravgargcs** [reply]
problem resolved thanks

**Comment by MDaskalovAtBE** [comment]
Hmm, it is a bit weird that if I use xlsx extension it works as the accepted answer suggests, but if I use identical code but just change it to CSV and the content of the file is empty so nothing happens, while the documentation suggests just swapping the extension should resolve in the same result. Any ideas how to make it work with CSV?

Works as expected -
```
        $writer = SimpleExcelWriter::streamDownload('your-export.xlsx')
            ->addRow([
                'first_name' => 'John',
                'last_name' => 'Doe',
            ])
            ->addRow([
                'first_name' => 'Jane',
                'last_name' => 'Doe',
            ]);

        return response()->streamDownload(function () use ($writer) {
            $writer->close();
        },  'your-export.xlsx');
```

Does not work - 

```
        $writer = SimpleExcelWriter::streamDownload('your-export.csv')
            ->addRow([
                'first_name' => 'John',
                'last_name' => 'Doe',
            ])
            ->addRow([
                'first_name' => 'Jane',
                'last_name' => 'Doe',
            ]);

        return response()->streamDownload(function () use ($writer) {
            $writer->close();
        },  'your-export.csv');
```

**Reply by t-prod** [reply]
Hi,

I confirm the download with CSV extension does not work. With .XLSX no problem 👍 
It's a shame to use an other package for CSV maybe pure PHP
Nobody have a solution ?

**Reply by hermeslin** [reply]
Hi, 
Moving the `$writer` into the `streamDownload` callback function might resolve the issue where the CSV extension doesn't work in normal situations.

```php
return response()->streamDownload(function () {
    $writer = SimpleExcelWriter::streamDownload('your-export.csv')
        ->addRow([
            'first_name' => 'John',
            'last_name' => 'Doe',
        ])
        ->addRow([
            'first_name' => 'Jane',
            'last_name' => 'Doe',
        ]);
    $writer->close();
},  'your-export.csv');
```
